### DISCUSSION TITLE
Does quarkus has native support to write reactive code in filter ?

### DISCUSSION THREAD

**Discussion by nitinty** [quarkusio/quarkus#43598] [Q&A]
Hi Team,

Does Quarkus natively support writing reactive code in filters? Upon checking, I noticed that the ResteasyReactiveContainerRequestFilter and ResteasyReactiveContainerResponseFilter interfaces do not provide methods that return a Uni<Void>.

Could you please confirm if I'm missing something or if there’s an alternative approach?

Thank you in advance for your assistance.

**Comment by nitinty** [comment]
@Ladicek  @geoand

**Comment by Ladicek** [comment]
The existence of methods `ResteasyReactiveContainerRequestContext.suspend()` and `resume()` suggests that this should be possible, but I'm no expert on RESTEasy Reactive.

**Reply by geoand** [reply]
Yes, that's the point

**Reply by nitinty** [reply]
@Ladicek @geoand 
Can you please tag someone who can answer or add any example here ?

**Reply by geoand** [reply]
Although you can implement `ResteasyReactiveContainerRequestFilter` it's much easier to use `@ServerRequestFilter` and return a `Uni`

**Reply by nitinty** [reply]
@geoand 

Can you please confirm if you're suggesting the following approach ?

```java
@Provider
public class AsyncServerRequestFilter {

    @ServerRequestFilter
    public Uni<Void> filter(ContainerRequestContext requestContext) {
        System.out.println("AsyncServerRequestFilter called in thread = "+ Thread.currentThread().getName());
        // Perform asynchronous operation and return a Uni
        return performAsyncOperation().onItem().transformToUni(result -> {
            System.out.println("Async operation result: " + result);
            // Return Uni<Void> to indicate the filter has completed
            return Uni.createFrom().voidItem();
        }).onFailure().invoke(failure -> {
            // Handle any errors that occurred during the async operation
            System.err.println("Failed async operation: " + failure.getMessage());
        });
    }

    private Uni<String> performAsyncOperation() {
        // Simulating an async task, like a non-blocking database call or external service call
        return Uni.createFrom().item("Async Operation Completed")
                  .onItem().delayIt().by(java.time.Duration.ofSeconds(2)); // Simulated delay
    }
}
```

**Reply by geoand** [reply]
In the code above you don't need `@Provider`.

**Reply by nitinty** [reply]
> Although you can implement `ResteasyReactiveContainerRequestFilter` it's much easier to use `@ServerRequestFilter` and return a `Uni`


@geoand 
I'm a bit confused about the first part of this answer. If I implement ResteasyReactiveContainerRequestFilter, Will the code execute in a reactive manner automatically ?
